test/parallel/test-promises-unhandled-rejections.js

```diff
@@ -0,0 +1,591 @@
+  });
+  var _reject;
+  var promise = new Promise(function(_, reject) {
+    _reject = reject;
+  });
+  setImmediate(function() {
+    promise.then(assert.fail, function(){});
+  });
+  _reject(e);
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var e2 = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e2, reason);
+    assert.strictEqual(promise2, promise);
+  });
+  var promise2 = Promise.reject(e).then(assert.fail, function(reason) {
+    assert.strictEqual(e, reason);
+    throw e2;
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var e2 = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+    assert.strictEqual(promise, promise);
+  });
+  var promise = Promise.reject(e);
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var e2 = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e2, reason);
+    assert.strictEqual(promise2, promise);
+  });
+  var promise2 = new Promise(function(_, reject) {
+    setTimeout(function() {
+      reject(e);
+    }, 1);
+  }).then(assert.fail, function(reason) {
+    assert.strictEqual(e, reason);
+    throw e2;
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var e2 = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e2, reason);
+    assert.strictEqual(promise2, promise);
+  });
+  var promise = new Promise(function(_, reject) {
+    setTimeout(function() {
+      reject(e);
+      process.nextTick(function() {
+        promise2 = promise.then(assert.fail, function(reason) {
+          assert.strictEqual(e, reason);
+          throw e2;
+        });
+      });
+    }, 1);
+  });
+  var promise2;
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.reject(e).then(assert.fail, function(){});
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  new Promise(function(_, reject) {
+    reject(e);
+  }).then(assert.fail, function(){});
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  var promise = Promise.reject(e);
+  process.nextTick(function() {
+    promise.then(assert.fail, function(){});
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  var promise = new Promise(function(_, reject) {
+    reject(e);
+  });
+  process.nextTick(function() {
+    promise.then(assert.fail, function(){});
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  var promise = new Promise(function(_, reject) {
+    reject(e);
+  });
+  process.nextTick(function() {
+    promise.then(assert.fail, function(){});
+  });
+});
+
+// State adapation tests
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.resolve().then(function() {
```

Comment:
catching a promise which is asynchronously rejected (via resolution to an asynchronously-rejected promise) prevents unhandledRejection